### ISSUE TITLE
如何设置多目标检测类别损失权重解决数据集分布不均衡的问题

### ISSUE THREAD

**Issue by wangyang581** [ultralytics/ultralytics#15615]
### Search before asking

- [X] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

我正在训练一个多目标检测的数据集，在我的数据集中，不同类别的标注框数量很不均衡，某些类别标注框数量比例达到了10:1，我想通过调整不同类别的损失权重训练模型优化检测效果，我参考了[https://github.com/ultralytics/ultralytics/issues/13330](url)中的设置方法，但是训练并没有成功出现了报错如下：
![image](https://github.com/user-attachments/assets/02a91602-b691-4eb2-882e-d84987762bf4)
我通过终端命令行进行训练，训练命令为：
yolo detect train data=custum.yaml model=yolov8s.pt epochs=500 imgsz=1280 device=3 batch=128 name=gamble_v3.4_s weight_decay=0.001 cls=[0.08,0.16,0.16,0.16,0.16,0.08,0.08,0.08]
看起来cls参数设置的有问题，请问我如何设置才能实现我想要的效果。

**Comment by glenn-jocher** [comment]
@wangyang581 to address the class imbalance issue, you can adjust the `cls` loss weights in your training configuration file. Ensure the `cls` parameter is correctly formatted as a list of floats corresponding to each class. For example, `cls=[0.08, 0.16, 0.16, 0.16, 0.16, 0.08, 0.08, 0.08]`. If the error persists, please verify if the issue continues with the latest package version.

**Comment by glenn-jocher** [comment]
@wangyang581 `cls` is the classification loss gain. This must be a float like 1.0, it can not be a list.

**Comment by chuang091** [comment]
@wangyang581 根據 #13330 你應該要調整的是 `custum.yaml` 裡面的weights 但你調整的這個list看起來像是**超參數 (hyperparameter)**
對於超參數， @glenn-jocher 所提到的 `cls` 請參考 #10375

**Comment by glenn-jocher** [comment]
@chuang091 您需要在 `custum.yaml` 文件中调整类别权重，而不是在命令行中设置 `cls` 参数。请参考 #10375 中关于 `cls` 的说明来正确设置超参数。

**Comment by hhhFuture** [comment]
> 您需要在 文件中调整类别权重，而不是在命令行中设置 参数。请参考 [#10375](https://github.com/ultralytics/ultralytics/issues/10375) 中关于 的说明来正确设置超参数。`custum.yaml``cls``cls`



> 您需要在 文件中调整类别权重，而不是在命令行中设置 参数。请参考 [#10375](https://github.com/ultralytics/ultralytics/issues/10375) 中关于 的说明来正确设置超参数。`custum.yaml``cls``cls`


Hello, I have reviewed what you mentioned in your previous reply

https://github.com/ultralytics/ultralytics/issues/13330

https://github.com/ultralytics/ultralytics/issues/10375. Coincidentally, in the current task, I have also encountered some problems that require setting the cls parameters, but I'm not sure if my setting is correct

The screenshot of the Settings is as follows:

![Image](https://github.com/user-attachments/assets/832133c1-f4c6-42dd-b027-12e70ae9a5d0)


My task is to detect the scissors gesture, but it misrecognizes other gestures. I tried to add a background image, but it still misrecognizes. After I reviewed a lot of relevant materials, a friend mentioned that other gestures could be labeled as another category and then a multi-object detection model could be trained. I wanted to try to do it this way. But I'm quite puzzled about this. If all the other gestures are trained as "other", I wonder if the model can fit them well, because there are still quite a lot of differences between gestures. Of course, if trained in this way, as long as the model can recognize the scissors gesture, it doesn't matter if it doesn't recognize other types of gestures well. After all, this was done in the first place only to reduce false recognition. Looking forward to your reply. Thank you in advance!

**Comment by hhhFuture** [comment]
> 您需要在 文件中调整类别权重，而不是在命令行中设置 参数。请参考 [#10375](https://github.com/ultralytics/ultralytics/issues/10375) 中关于 的说明来正确设置超参数。`custum.yaml``cls``cls`

Hello, I have reviewed what you mentioned in your previous reply
https://github.com/ultralytics/ultralytics/issues/13330
https://github.com/ultralytics/ultralytics/issues/10375. Coincidentally, in the current task, I have also encountered some problems that require setting the cls parameters, but I'm not sure if my setting is correct
The screenshot of the Settings is as follows:

![Image](https://github.com/user-attachments/assets/72ff2322-3bfe-4c7a-8667-2395b413c3e3)

The screenshots of the training are as follows
from ultralytics import YOLO
import time

if __name__ == '__main__':
    model = YOLO("yolov8n.pt")
    model.train(
        data=r"E:\Project_Gesture\model_script\scissors.yaml",
        imgsz=640,
        device=0,
        lr0=0.005, 
        epochs=200,
        batch=64,
        close_mosaic=10,
        name="3clss_sciss_yolo8n",
        fliplr=0.5,  
        flipud=0.5, 
        degrees=15, 
        mosaic=0.5,  
        mixup=0.3, 
        plots=True,  
        scale=0.2,  
    )

My task is to detect the scissors gesture, but it will misrecognize other gestures. I tried to add the background image, but it still misrecognizes. Moreover, as the number of added gestures increases, the model's effect of positive recognition will also decrease. The model will also misrecognize gestures from a distance, whether they are scissors gestures or not. After I reviewed a lot of relevant materials, a friend mentioned that other gestures could be labeled as another category and then a multi-object detection model could be trained. I wanted to try to do it this way. But I'm quite puzzled about this. If all the other gestures are trained as "other", I wonder if the model can fit them well, because there are still quite a lot of differences between gestures. Of course, if trained in this way, as long as the model can recognize the scissors gesture, it doesn't matter if it doesn't recognize other types of gestures well. After all, this was done in the first place only to reduce false recognition. Looking forward to your reply. Thank you in advance!

**Comment by glenn-jocher** [comment]
@hhhFuture Your approach of adding other gestures as an "other" class is reasonable for reducing false positives. For class weights to handle imbalance, you need to add a `class_weights` parameter to your dataset YAML file, not in the training code. The `cls` parameter you mentioned is the overall classification loss weight (should remain a single float like 0.5), not per-class weights. In your `scissors.yaml` file, add something like `class_weights: [1.0, 0.5]` where the values correspond to weights for each class in order.

**Comment by hhhFuture** [comment]
> 您将其他手势添加为“其他”类的方法是合理的，以减少误报。要使类权重处理不平衡，您需要向数据集 YAML 文件添加参数，而不是在训练代码中添加参数。您提到的参数是整体分类损失权重（应保持单个浮点数，如 0.5），而不是每个类别的权重。在您的文件中，添加类似 where the values to weights for each class （按顺序） 对应的内容。`class_weights``cls``scissors.yaml``class_weights: [1.0, 0.5]`

Thank you for your reply. scissors.yaml is the YAML file of the dataset. I have set the cls parameter in this file. So, does the cls parameter only need to be set in the YAML of the dataset? The classification loss weights in the total loss function can be adjusted without modifying other codes, right?
CLS Settings I are examples of reference you provide in https://github.com/ultralytics/ultralytics/issues/13330, wrote after the code is as follows:
# Example: Class weights calculation
class_counts = [30, 130]  # Number of images per class
total_samples = sum(class_counts)
class_weights = [total_samples / count for count in class_counts]

print(class_weights) # Output: [5.0, 1.1538461538461537]
The screenshot is as follows:

![Image](https://github.com/user-attachments/assets/eae53f1d-d605-4302-96ff-39f054a335a7)

However, in the above answer, you mentioned that "a single floating-point number, such as 0.5, should be maintained", and it was mentioned in the ultralytics documentation that its range is from 0.2 to 0.4. So I'm very confused now. Which one is the right one, or which one is the reasonable way?

**Comment by hhhFuture** [comment]
> 您将其他手势添加为“其他”类的方法是合理的，以减少误报。要使类权重处理不平衡，您需要向数据集 YAML 文件添加参数，而不是在训练代码中添加参数。您提到的参数是整体分类损失权重（应保持单个浮点数，如 0.5），而不是每个类别的权重。在您的文件中，添加类似 where the values to weights for each class （按顺序） 对应的内容。`class_weights``cls``scissors.yaml``class_weights: [1.0, 0.5]`


Thank you for your reply. scissors.yaml is the YAML file of the dataset. I have set the cls parameter in this file. So, does the cls parameter only need to be set in the YAML of the dataset?
I refer to your example set the CLS parameters, the corresponding link https://github.com/ultralytics/ultralytics/issues/13330 here in this case, you provide examples of code:
~~~python
# Example: Class weights calculation
class_counts = [30, 130]  # Number of images per class
total_samples = sum(class_counts)
class_weights = [total_samples / count for count in class_counts]
print(class_weights) # Output: [5.0, 1.1538461538461537]
~~~
The screenshot is as follows:

![Image](https://github.com/user-attachments/assets/d5c66abc-2a11-47e8-a91b-d905e4cc3a87)

However, in the above answer, you mentioned that "should remain a single float like 0.5", and it was mentioned in the ultralytics documentation that its range is from 0.2 to 0.4. So I'm very confused now. Which one is the right one, or which one is the reasonable way? In my task, in order to solve the misrecognition of other gestures, I classified the other gestures as the "other" class. There are approximately 90,000 copies in the "other" category, 60,000 copies in Forward_scissor and 50,000 copies in Reverse_scissor respectively. As long as the model can recognize the scissors gesture, that's fine. So I want to set the cls to allocate the classification weights for each target. This is also the reason why it's so troublesome. In my task, how should I set its parameters better?

**Comment by glenn-jocher** [comment]
@hhhFuture You're confusing two different parameters. The `cls` parameter is the overall classification loss weight (single float, typically 0.5) that balances classification loss against box loss and other components - this goes in training arguments, not your dataset YAML. For handling class imbalance, you need `class_weights` in your dataset YAML file. Based on your counts (90k "other", 60k forward, 50k reverse), calculate inverse frequency weights: `class_weights: [0.67, 1.0, 1.2]` and add this line to your `scissors.yaml` file.

**Comment by hhhFuture** [comment]
Thank you very much for your reply. I went back to check the hyperparameters on the ultralytics official website again. As you said, I was indeed confused. Thank you for your reminder. I'd like to share my own understanding and the current problems. I hope you can take a look. Thank you for your hard work.
My understanding is as follows:
The loss function of yolo is composed of cls loss,box loss and dfl. In the training, if the cls is set in model.train, the higher weight emphasizes the correct category prediction. In the document, I found that the cls value range is (0.2, 0.4), and the box value range is (0.02, 0.2). When training the model, an args.yaml is generated. What puzzles me is that I saw the box in args.yaml set to 7.5, cls:0.5,dfl:1.5. So, can I understand that the parameters set in args.yaml are the default parameters? model.train() can emphasize the weight of a certain loss by setting the values of cls and box and then performing operations with the values in args.yaml?
The screenshot is as follows:

![Image](https://github.com/user-attachments/assets/80de7348-c874-4610-ad83-ff3278894d23)

For the data imbalance of the processing category, Settings need to be made in the yaml dataset YAML file. I don't know if it is set like this. How did you calculate [0.67, 1.0, 1.2]? I'm really looking forward to your reply. Thank you

![Image](https://github.com/user-attachments/assets/d3fe3c9f-9b56-493e-8410-c1f914ff2cb7)

~~~python
train: [E:\All_Data\Gesture_Data\train.txt]
val: [E:\All_Data\Gesture_Data\val.txt]

cls class_weights class_weights: [0.67, 1.0, 1.2]

nc: 3

names: ['Forward_scissor', 'Reverse_scissor','other']
~~~

**Comment by glenn-jocher** [comment]
@hhhFuture The class weights [0.67, 1.0, 1.2] are calculated using your middle class (60k) as reference: 60k/90k=0.67, 60k/60k=1.0, 60k/50k=1.2. This gives higher weights to underrepresented classes.

Regarding loss components: the `cls=0.5`, `box=7.5`, `dfl=1.5` values in args.yaml are the default loss component weights that balance different parts of the total loss function - these are separate from class imbalance weights.

Your dataset YAML syntax needs fixing. Remove the extra text and use:

```yaml
train: E:\All_Data\Gesture_Data\train.txt
val: E:\All_Data\Gesture_Data\val.txt

class_weights: [0.67, 1.0, 1.2]

nc: 3
names: ['Forward_scissor', 'Reverse_scissor', 'other']
```

**Comment by hhhFuture** [comment]
> 类权重 [0.67， 1.0， 1.2] 是以您的中产阶级 （60k） 为参考计算的：60k/90k=0.67、60k/60k=1.0、60k/50k=1.2。这为代表性不足的类别提供了更高的权重。
> 
> 关于损失分量：args.yaml 中的 ， ， 值是平衡总损失函数不同部分的默认损失分量权重 - 这些与类不平衡权重是分开的。`cls=0.5``box=7.5``dfl=1.5`
> 
> 您的数据集 YAML 语法需要修复。删除多余的文本并使用：
> 
> train: E:\All_Data\Gesture_Data\train.txt
> val: E:\All_Data\Gesture_Data\val.txt
> 
> class_weights: [0.67, 1.0, 1.2]
> 
> nc: 3
> names: ['Forward_scissor', 'Reverse_scissor', 'other']

Thank you for your reply. Thank you for your patient answer to my question. Indeed, the code can run normally in your way. I still have a question that I'd like to consult you. I'm sorry to have asked you so many questions. I'm very sorry.
The questions are as follows:
If I want the model to pay more attention to the accuracy of classification, can I set the cls in model.train()? I tried it myself. If it is set, the cls of args.yaml will change. If the cls of model.train() is set to 1.0, the cls of args.yaml will also become 1.0. The reason why I hope the model classification is so accurate is that when using yolov8 to perform the scissors gesture now, false recognition is too difficult to remove. If too much background is added, the formal recognition will drop, and if too little is added, the false recognition cannot be removed. I tried to classify the misidentified gestures into the "other" category. Even though the dataset was balanced, the model still found it much easier to classify the gestures into the "other" category, which was very frustrating! I am still very grateful for your repeated teachings. I believe that as more and more people like you come, the community will become better and better!

**Comment by glenn-jocher** [comment]
@hhhFuture Yes, you can set `cls=1.0` in `model.train()` to increase the classification loss weight and make the model focus more on classification accuracy. This will indeed override the default value in args.yaml. For your gesture recognition problem with false positives, increasing the classification weight is a reasonable approach to try - you could even experiment with higher values like `cls=1.5` or `cls=2.0` to further emphasize correct classification over bounding box precision.

**Comment by Y-T-G** [comment]
Duplicate of https://github.com/ultralytics/ultralytics/issues/21082

**Comment by CVRS-CJH** [comment]
我看你train设置的是txt来传入图像的绝对路径，请问训练数据集的label你是放在哪里呀，我这边一直提示找不到label文件。

**Comment by hhhFuture** [comment]
> 我看你train设置的是txt来传入图像的绝对路径，请问训练数据集的label你是放在哪里呀，我这边一直提示找不到label文件。
Hello, If I remember correctly, for yolo training, it is indeed only necessary to pass in the txt of the images, and the model finds the labels by replacing the images characters in the path.
For example, the content of train.txt:
E:\xxx\xxx\xxx\images\first.jpg
yolo will find the labels by replacing "images" with "labels"

**Comment by glenn-jocher** [comment]
@CVRS-CJH @hhhFuture is correct. YOLO automatically finds labels by replacing "images" with "labels" in the path. Ensure your dataset structure follows this convention:

```
dataset/
├── images/
│   ├── train/
│   └── val/
└── labels/
    ├── train/
    └── val/
```

If your images are at `E:\xxx\images\first.jpg`, the corresponding label should be at `E:\xxx\labels\first.txt`.

**Comment by Chenniubi1** [comment]
> 班级权重 [0.67， 1.0， 1.2] 以您的中产阶级 （60k） 为参考计算：60k/90k=0.67、60k/60k=1.0、60k/50k=1.2。这为代表性不足的阶层提供了更高的权重。
> 
> 关于损失分量：args.yaml 中的 ， ， 值是平衡总损失函数不同部分的默认损失分量权重 - 这些与类不平衡权重是分开的。`cls=0.5``box=7.5``dfl=1.5`
> 
> 数据集 YAML 语法需要修复。删除多余的文本并使用：
> 
> train: E:\All_Data\Gesture_Data\train.txt
> val: E:\All_Data\Gesture_Data\val.txt
> 
> class_weights: [0.67, 1.0, 1.2]
> 
> nc: 3
> names: ['Forward_scissor', 'Reverse_scissor', 'other']

您好，您提到的直接在yaml中新增class_weights参数，但是我详细阅读了yolo代码发现这个参数根本就没有传递进去，是一个无效的参数

**Comment by glenn-jocher** [comment]
@Chenniubi1 您说得对，`class_weights` 参数确实不是 YOLO 中的有效参数。我为之前的错误信息道歉。

对于类别不平衡问题，YOLO 目前不支持在 YAML 文件中直接设置每个类别的权重。您可以尝试以下方法：

1. 调整 `cls` 参数来增加分类损失的整体权重
2. 使用数据增强和重采样策略来平衡数据集
3. 考虑使用 Focal Loss 等专门处理类别不平衡的损失函数

感谢您指出这个错误！
